### ISSUE TITLE
Speed of restrict_chars

### ISSUE THREAD

**Issue by jake-87** [StavromulaBeta/cognate#14]
Due to potentially being called a lot in large programs, the speed of restrict_chars could be extremely important. Is there a reason for the replacement words, here:
```
	char* with[]	 = {"DASH", "XMARK", "QMARK", "EQ", "LT", "GT", "PLUS", "STAR", "SLASH"};
```
to be full strings, instead of say, `'D', 'X', 'Q', 'E', 'L', 'G', 'P', 'S', 'H'`?
The only thing I can think of is debugging.
If these were to be replaced with single char versions, you'd be able to much more quickly remove them. 

Note this will not affect other variables, due to the sanitation process - I can provide examples if needed.

We could always provide two versions, using a flag like `DEBUG` to enable the slower version.

**Comment by jake-87** [comment]
Currently restrict_chars is ~`O(n^4)` - I believe it can be brought down to `O(n)` using single char replacements.

**Comment by StavromulaBeta** [comment]
I think the current version is still O(n) since `replace` and `with` have a small fixed size. However I agree that your suggestion is an improvement. A lot of the functions in `src/runtime.c` would have to be renamed though.

**Comment by StavromulaBeta** [comment]
I think we could also replace the `calloc` with `strdup`

**Comment by jake-87** [comment]
Yes, strdup should work. I always tend to forget it exists. I’ll have a look at runtime.c and see what can be done on that front - I might make a draft PR to track work, if you think this is a good idea.

**Comment by StavromulaBeta** [comment]
sounds great!

**Comment by jake-87** [comment]
fixed by #15
